cpp/src/arrow/dataset/file_csv.cc

```diff
@@ -105,18 +106,33 @@ static inline Result<std::shared_ptr<csv::StreamingReader>> OpenReader(
     const FileSource& source, const CsvFileFormat& format,
     const std::shared_ptr<ScanOptions>& scan_options = nullptr,
     MemoryPool* pool = default_memory_pool()) {
-  ARROW_ASSIGN_OR_RAISE(auto input, source.Open());
-
   auto reader_options = GetReadOptions(format);
-  ARROW_ASSIGN_OR_RAISE(auto first_block, input->ReadAt(0, reader_options.block_size));
-  RETURN_NOT_OK(input->Seek(0));
 
-  const auto& parse_options = format.parse_options;
+  std::shared_ptr<io::InputStream> input;
+  // If applicable, the buffer that backs first_block
+  std::shared_ptr<Buffer> first_block_buf;
+  util::string_view first_block;
+  ARROW_ASSIGN_OR_RAISE(auto file, source.Open());
+  if (format.compression == Compression::UNCOMPRESSED) {
+    input = file;
```

Comment:
(For follow up, maybe ARROW-8981) maybe it'd be more useful to encapsulate compressed FIleSources with an overload of Open()
```suggestion
  ARROW_ASSIGN_OR_RAISE(std::shared_ptr<io::InputStream> file, source.Open(format.compression));
```
rather than the (currently ignored) FileSource::compression property